bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/StateManagerTest.java

```diff
@@ -0,0 +1,207 @@
+    public void setUp() throws Exception {
+        super.setUp();
+        zkUtil.createBKEnsemble("/" + runtime.getMethodName());
+        rm = new MockZKRegistrationManager();
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        if (rm != null) {
+            rm.close();
+        }
+    }
+
+    private static class MockZKRegistrationManager extends ZKRegistrationManager {
+        boolean registerFailed = false;
+        void setRegisterFail(boolean failOrNot){
+            registerFailed = failOrNot;
+        }
+        @Override
+        public void registerBookie(String bookieId, boolean readOnly) throws BookieException {
+            if (registerFailed) {
+                throw BookieException.create(-100);
+            }
+            super.registerBookie(bookieId, readOnly);
+        }
+
+    }
+
+    /**
+     * Bookie should shutdown when it register to Registration service fail.
+     * On ZooKeeper exception, should return exit code ZK_REG_FAIL = 4
+     */
+    @Test
+    public void testShutdown() throws Exception {
+        File tmpDir = createTempDir("stateManger", "test");
+
+        final ServerConfiguration conf = TestBKConfiguration.newServerConfiguration();
+        conf.setJournalDirName(tmpDir.getPath())
+            .setLedgerDirNames(new String[] { tmpDir.getPath() })
+            .setZkServers(zkUtil.getZooKeeperConnectString());
+
+        BookieServer bkServer = new BookieServer(conf) {
+            protected Bookie newBookie(ServerConfiguration conf)
+                    throws IOException, KeeperException, InterruptedException,
+                    BookieException {
+                Bookie bookie = new Bookie(conf);
+                rm.setRegisterFail(true);
+                rm.initialize(conf, () -> {}, NullStatsLogger.INSTANCE);
+                LOG.info(" is {} ", zkc == null);
+                bookie.setRegistrationManager(rm);
+                return bookie;
+            }
+        };
+        bkServer.start();
+        bkServer.join();
+        assertTrue("Failed to return failCode ZK_REG_FAIL",
+                ExitCode.ZK_REG_FAIL == bkServer.getExitCode());
+    }
+
+    /**
+     * StateManager can transition between writable mode and readOnly mode if it was not created with readOnly mode.
+     */
+    @Test
+    public void testBookieTransitions() throws Exception {
+        File tmpDir = createTempDir("stateManger", "test");
+
+        final ServerConfiguration conf = TestBKConfiguration.newServerConfiguration();
+        conf.setJournalDirName(tmpDir.getPath())
+                .setLedgerDirNames(new String[] { tmpDir.getPath() })
+                .setZkServers(zkUtil.getZooKeeperConnectString());
+        Bookie bookie = new Bookie(conf);
+        bookie.start();
+        assertTrue(bookie.isRunning());
+
+        bookie.getStateManager().transitionToReadOnlyMode();
+        //sleep a little to wait transition finish
+        Thread.sleep(1000);
```

Comment:
This is not a good idea. It is not good to use `sleep(1000)`. Can you wait on the future returned by transitionToReadOnlyMode?